source/channel_bootstrap.c

```diff
@@ -802,6 +808,8 @@ static inline int s_new_client_channel(
         memcpy(endpoint.address, host_name, strlen(host_name));
         endpoint.port = 0;
 
+        s_connection_args_acquire(client_connection_args);
```

Comment:
you moved this too high.
it should go immediately before the call to aws_socket_connect(), and if the connect call fails you should s_connection_args_release() before the goto error;